[SPARK-19357][ML] Adding parallel model evaluation in ML tuning

BryanCutler opened the pull request:
## What changes were proposed in this pull request?
Modified `CrossValidator` and `TrainValidationSplit` to be able to evaluate models in parallel for a given parameter grid.  The level of parallelism is controlled by a parameter `numParallelEval` used to schedule a number of models to be trained/evaluated so that the jobs can be run concurrently.  This is a naive approach that does not check the cluster for needed resources, so care must be taken by the user to tune the parameter appropriately.  The default value is `1` which will train/evaluate in serial. 

## How was this patch tested?
Added unit tests for CrossValidator and TrainValidationSplit to verify that model selection is the same when run in serial vs parallel.  Manual testing to verify tasks run in parallel when param is > 1. Added parameter usage to relevant examples.

BryanCutler committed:
Changed CrossValidator and TrainValidationSplit fit methods to evaluate models in paralell

BryanCutler commented:
CC @mlnick, this doesn't yet address the issue of double caching, but I wanted to post what I had so far to discuss the additional param and default value.  I'll be looking into the caching next.

SparkQA commented:
**[Test build #72267 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72267/testReport)** for PR 16774 at commit [`5650e98`](https://github.com/apache/spark/commit/5650e98a580544303dc1185568be992d9304707a).

SparkQA commented:
**[Test build #72267 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72267/testReport)** for PR 16774 at commit [`5650e98`](https://github.com/apache/spark/commit/5650e98a580544303dc1185568be992d9304707a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72267/
Test PASSed.

MLnick commented on mllib/src/main/scala/org/apache/spark/ml/tuning/CrossValidator.scala:
```diff
@@ -100,31 +104,44 @@ class CrossValidator @Since("1.2.0") (@Since("1.4.0") override val uid: String)
+
+      // Evaluate models concurrently, limited by using a sliding window over models
+      val foldMetrics = models.zip(epm).grouped(numPar).map { win =>
+        win.par.map { m =>
```
I'd like to use a `case` here to be more clear, so `.map { case (model, paramMap) =>` or similar.

MLnick commented on mllib/src/main/scala/org/apache/spark/ml/tuning/CrossValidator.scala:
```diff
@@ -100,31 +104,44 @@ class CrossValidator @Since("1.2.0") (@Since("1.4.0") override val uid: String)
+
+      // Fit models concurrently, limited by using a sliding window over models
+      val models = epm.grouped(numPar).map { win =>
+        win.par.map(est.fit(trainingDataset, _))
```
I'd prefer to be a bit more verbose here and use a closure variable like `paramMap` (or `epm` to follow existing impl) instead of `_` just be clear & more readable.

MLnick commented on mllib/src/main/scala/org/apache/spark/ml/tuning/TrainValidationSplit.scala:
```diff
@@ -106,18 +110,21 @@ class TrainValidationSplit @Since("1.5.0") (@Since("1.5.0") override val uid: St
-    val models = est.fit(trainingDataset, epm).asInstanceOf[Seq[Model[_]]]
+    // Fit models concurrently, limited by using a sliding window over models
+    val models = epm.grouped(numPar).map { win =>
+      win.par.map(est.fit(trainingDataset, _))
```
Same comment as for cross-val applies here

MLnick commented on mllib/src/main/scala/org/apache/spark/ml/tuning/TrainValidationSplit.scala:
```diff
@@ -106,18 +110,21 @@ class TrainValidationSplit @Since("1.5.0") (@Since("1.5.0") override val uid: St
-    }
+    // Evaluate models concurrently, limited by using a sliding window over models
+    val metrics = models.zip(epm).grouped(numPar).map { win =>
+      win.par.map { m =>
```
Same comment as for cross-val applies here

MLnick commented on mllib/src/main/scala/org/apache/spark/ml/tuning/TrainValidationSplit.scala:
```diff
@@ -106,18 +110,21 @@ class TrainValidationSplit @Since("1.5.0") (@Since("1.5.0") override val uid: St
-    // multi-model training
     logDebug(s"Train split with multiple sets of parameters.")
-    val models = est.fit(trainingDataset, epm).asInstanceOf[Seq[Model[_]]]
+    // Fit models concurrently, limited by using a sliding window over models
```
perhaps add "sliding window of size 'numPar' over ..." - can add to all instances of this comment

MLnick commented on mllib/src/main/scala/org/apache/spark/ml/tuning/CrossValidator.scala:
```diff
@@ -51,7 +51,7 @@ private[ml] trait CrossValidatorParams extends ValidatorParams {
   def getNumFolds: Int = $(numFolds)
 
-  setDefault(numFolds -> 3)
+  setDefault(numFolds -> 3, numParallelEval -> 1)
```
May make more sense to put the `setDefault` call in the parent trait `ValidatorParams`

MLnick commented on mllib/src/main/scala/org/apache/spark/ml/tuning/ValidatorParams.scala:
```diff
@@ -67,6 +67,17 @@ private[ml] trait ValidatorParams extends HasSeed with Params {
+   * @group param
+   */
+  val numParallelEval: IntParam = new IntParam(this, "numParallelEval",
+    "max number of models to evaluate in parallel, 1 for serial evaluation")
```
Add note about default value being `1`

MLnick commented on mllib/src/main/scala/org/apache/spark/ml/tuning/CrossValidator.scala:
```diff
@@ -100,31 +104,44 @@ class CrossValidator @Since("1.2.0") (@Since("1.4.0") override val uid: String)
     instr.logParams(numFolds, seed)
     logTuningParams(instr)
 
+    // Compute metrics for each model over each fold
```
technically it's each `estimator` rather than `model` - since both the fitting (estimator or pipeline) and evaluation (model or pipeline model) is in parallel.

BryanCutler replied:
Hmm, I was just referencing the general process that a metric is computed by evaluating the model predictions on the validation set, and that this is done at each fold.  I don't think that has any bearing to the concurrency here, but I added this just because the process is a little more complicated with the addition of a sliding window...

MLnick reviewed (changes requested):
Made a first pass.

Let's add unit tests in both test suites to check that the metric results for default `numPar` are the same as for `numPar > 1`.

It would be good to also add a section about the parallel execution option to each validator's doc page.

I think it's important to note in the docs that this will actually only really work as expected if the `FAIR` scheduler is enabled, otherwise I don't think things will actually be executed concurrently.

BryanCutler commented:
Thanks for the review @MLnick!  All good suggestions, I'll work on an update and addition of the unit test.

BryanCutler commented:
Regarding the default value of the new parameter, do you think `1`, so that it is run in serial as it is currently doing, is the best option?  It's definitely safe, but I wonder if most people will even know that the parameter exists..

Do you know of any possible way to maybe automatically choose a value?  I remember a while ago I saw the mllib ALS algorithm automatically choose the block size, but I don't think that's done in the ml version.  Maybe there were too many problems doing this..

BryanCutler committed:
made closure vars more explicit, moved param default to trait

BryanCutler committed:
added paramvalidator for numParallelEval to ensure >=1

BryanCutler committed:
added test cases for CrossValidation and TrainValidationSplit

BryanCutler committed:
added numParallelEval param usage to examples

MLnick commented:
I'd say coming up with a heuristic or algorithm to automatically set the parallel execution param is going to be pretty challenging, since it depends on the details of the individual pipeline components, cluster resources, etc. So for now let's leave it fixed.

I think setting the default to match current behavior is best (i.e. `1`). We can add a section to the docs explaining its effect, and also add a couple lines to each example to illustrate typical usage.

BryanCutler committed:
added documentation to ml-tuning

BryanCutler committed:
changed sliding window limit to use a semaphore instead to prevent waiting for entire window to finish

BryanCutler committed:
added note about parallelism capped by Scala collection thread pool, adjusted comments

SparkQA commented:
**[Test build #73021 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73021/testReport)** for PR 16774 at commit [`6a9b735`](https://github.com/apache/spark/commit/6a9b73585a0da0960f9a25be24e852e0b1c3e01d).

BryanCutler commented:
> I think setting the default to match current behavior is best (i.e. 1).

I agree, just wanted to bring it up in case others had differing view since it was a concern in the JIRA.

>I think it's important to note in the docs that this will actually only really work as expected if the FAIR scheduler is enabled, otherwise I don't think things will actually be executed concurrently.

I ran some tests with and without the FAIR scheduler enabled.  Without FAIR scheduler, if there are enough resources in the cluster (e.g. cores) then the tasks are run concurrently.  If not enough resources, then the jobs will wait.  With FAIR scheduler the tasks are run concurrently and share the available cores.  Both will complete model selection in about the same time, with equal resources, so I think the main benefit of FAIR is for multi-users so that one users jobs don't get starved.

SparkQA commented:
**[Test build #73021 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73021/testReport)** for PR 16774 at commit [`6a9b735`](https://github.com/apache/spark/commit/6a9b73585a0da0960f9a25be24e852e0b1c3e01d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/73021/
Test PASSed.

BryanCutler commented:
@MLnick , I updated this PR:
* added unit tests
* added usage of parameter to examples
* updated ml-tuning.md documentation
* changed sliding window to a bounded semaphore to limit parallelism, this way as soon as one job is complete, another will get scheduled without waiting for the entire window to finish - it's a bit cleaner too

Please take a look when you get a chance, thanks!

thunterdb commented on mllib/src/test/scala/org/apache/spark/ml/tuning/CrossValidatorSuite.scala:
```diff
@@ -121,6 +121,33 @@ class CrossValidatorSuite
+      .setNumFolds(2)
+      .setNumParallelEval(1)
+    val cvSerialModel = cv.fit(dataset)
+    cv.setNumParallelEval(2)
```
this test is not deterministic now. @MLnick , do you know if we have some utilities to retry tests a much of times?

BryanCutler replied:
Yes, the models could now be evaluated in a different order, but the end result of returning the best model would be deterministic and shouldn't ever change.

MLnick replied:
Yes, I'd say we don't care about deterministic execution order here - we care about the result being the same regardless of execution order.

thunterdb commented:
Thanks for working on this task, this is a much requested feature. While it will work for simple cases in the current shape, it is going to cause some issues for any complex deployments (Apache Toree, Livy, Databricks, etc.) because the threadpool that controls the computations is not managed. The default assumption with `.par` is a lot of quick tasks. With the current implementation, because the same thread pool is going to be shared across all the parallel collections, users are going to encounter some mysterious freezes in other places, while the ML models are finishing to train (I am talking from experience here).

While the situation with `.par` has notably improved with scala 2.10, it is better to:
 - create a dedicated thread pool for each `.fit`, that users can replace.
 - use futures, of which the execution context is tied to the thread pool above.
 - not use semaphores, but instead rely on the thread limit at the thread pool level to cap the number of concurrent execution.

If you do not do that, users in a shared environment like any of the above will experience some mysterious freezes depending on what other users are doing. Ideally the default resources should be tied to `SparkSession`, but we can start with a default static pool marked as experimental API.

More concretely, the API should look like this, I believe:
```
  def setNumParallelEval(num: Int) // Creates an execution context with the given max number of threads
  def setExecutionExecutorService(exec: ExecutorService) // Will use the given executor service instead of an executor service shared by all the ML calculations
```

See the doc in:
https://twitter.github.io/scala_school/concurrency.html#executor
https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ExecutorService.html

More about parallel collections:
http://stackoverflow.com/questions/5424496/scala-parallel-collections-degree-of-parallelism
http://stackoverflow.com/questions/14214757/what-is-the-benefit-of-using-futures-over-parallel-collections-in-scala

BryanCutler commented:
Hi @thunterdb , thanks for the review and all of the details you provided!  I agree that a configurable execution service would be needed for running under a shared environment instead of simply using the default context.  I'll rework this and make use of Futures too.

BryanCutler committed:
reworked to use ExecutorService and Futures

SparkQA commented:
**[Test build #73545 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73545/testReport)** for PR 16774 at commit [`1c2e391`](https://github.com/apache/spark/commit/1c2e391e74696c60690f477fe4e0b636f0d98318).

BryanCutler committed:
fixed wildcard import

SparkQA commented:
**[Test build #73549 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73549/testReport)** for PR 16774 at commit [`9e055cd`](https://github.com/apache/spark/commit/9e055cd9e3c6fc507deb2db4981efe91b8be994a).

BryanCutler commented:
@thunterdb and @MLnick I updated this to use a configurable `ExecutorService` and `Future`s instead of Scala parallel collections.  The ExecutorService is retrieved by a function to lazily initialize the thread-pool until it is needed.  When running cross-validation there is a trade-off between waiting for the all calculations to complete at each split before continuing to the next and excessively caching all of the split datasets at once.  I chose to be safe and just wait for calculations at each split to complete, but let me know what you guys think - hopefully this makes sense :)

SparkQA commented:
**[Test build #73545 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73545/testReport)** for PR 16774 at commit [`1c2e391`](https://github.com/apache/spark/commit/1c2e391e74696c60690f477fe4e0b636f0d98318).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/73545/
Test PASSed.

SparkQA commented:
**[Test build #73549 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73549/testReport)** for PR 16774 at commit [`9e055cd`](https://github.com/apache/spark/commit/9e055cd9e3c6fc507deb2db4981efe91b8be994a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/73549/
Test PASSed.

MLnick commented on docs/ml-tuning.md:
```diff
@@ -55,6 +55,9 @@ for multiclass problems. The default metric used to choose the best `ParamMap` c
 method in each of these evaluators.
 
 To help construct the parameter grid, users can use the [`ParamGridBuilder`](api/scala/index.html#org.apache.spark.ml.tuning.ParamGridBuilder) utility.
+Sets of parameters from the parameter grid can be evaluated in parallel by setting `numParallelEval` with a value of 2 or more (a value of 1 will evaluate in serial) before running model selection with `CrossValidator` or `TrainValidationSplit`.
```
I think it's a little clearer to say something like "By default, sets of parameters ... will be evaluated in serial. Parameter evaluation can be done in parallel by setting ..."

MLnick replied:
We should add a note that this only works in Scala/Java, not Python (or R)

MLnick commented on docs/ml-tuning.md:
```diff
@@ -55,6 +55,9 @@ for multiclass problems. The default metric used to choose the best `ParamMap` c
 To help construct the parameter grid, users can use the [`ParamGridBuilder`](api/scala/index.html#org.apache.spark.ml.tuning.ParamGridBuilder) utility.
+Sets of parameters from the parameter grid can be evaluated in parallel by setting `numParallelEval` with a value of 2 or more (a value of 1 will evaluate in serial) before running model selection with `CrossValidator` or `TrainValidationSplit`.
+The value of `numParallelEval` should be chosen carefully to maximize parallelism without exceeding cluster resources, and will be capped at the number of cores in the driver system.  Generally speaking, a value up to 10 should be sufficient for most clusters.
+
```
Also will need to mention that custom `ExecutorService` can be specified, and some detail on the default thread pool it creates (and that it is a new separate pool to avoid blocking any of the default Scala pools).

BryanCutler replied:
Since that API is marked as experimental, maybe it would be better to not document right away until we are sure this is what we need?

MLnick commented on mllib/src/main/scala/org/apache/spark/ml/tuning/ValidatorParams.scala:
```diff
@@ -67,6 +71,39 @@ private[ml] trait ValidatorParams extends HasSeed with Params {
+   */
+  @Experimental
+  @InterfaceStability.Unstable
+  def setExecutorService(getExecutorService: (Int) => ExecutorService): Unit = {
```
I think I prefer a simpler API: 

```scala
 def setExecutorService(executorService: ExecutorService): Unit = {
```

Not sure the function version will work nicely with Java.

I think the idea should be that setting `numParallelEval` will specify the number of threads for the default thread-pool, while a custom one can be set with this method (advanced usage). If set, the custom one will override.

BryanCutler replied:
I think you are probably right about the Java thing, I'll test that out.

>custom one can be set with this method (advanced usage). If set, the custom one will override.

That's how this works, the executor service can be overridden and set any size thread pool, it is just passed the `numParallelEval` param for convenience in case it wants to use it.

BryanCutler replied:
One reason I did not do the simpler API that you mentioned is then it requires the ExecutorService to be instantiated with a default and then exists for the life of `CrossValidator` or `TrainValidationSplit`.  The way it is currently, by default the ExecutorService is created at the beginning of `fit` when it is needed, and then cleaned up at the end.

BryanCutler replied:
ah, I see what you are getting at.  Even if I override with a custom ExecutorService but leave `numParallelEval` to 1, it will not use the custom.. I can fix that

MLnick commented on mllib/src/main/scala/org/apache/spark/ml/tuning/ValidatorParams.scala:
```diff
@@ -67,6 +71,39 @@ private[ml] trait ValidatorParams extends HasSeed with Params {
+   *
+   * @group param
+   */
+  val numParallelEval: IntParam = new IntParam(this, "numParallelEval",
```
Perhaps set this as group `expertParam` for now

MLnick commented on mllib/src/main/scala/org/apache/spark/ml/tuning/ValidatorParams.scala:
```diff
@@ -67,6 +71,39 @@ private[ml] trait ValidatorParams extends HasSeed with Params {
+   * size of `numParallelEval`
+   */
+  protected var executorServiceFactory: (Int) => ExecutorService = {
+    (requestedMaxThreads: Int) => ThreadUtils.newDaemonCachedThreadPool(
```
I think it may be problematic (and probably unnecessary) to have daemon threads here - we don't have a shutdown hook.

We could add a shutdown hook to tie the lifecycle to SparkContext (SparkSession); perhaps in `setExecutorService`:

```scala
...
ShutdownHookManager.addShutdownHook(() => executorService.shutdown())
...
```

BryanCutler replied:
So my thinking was that if the thread calling fit is terminated, it would have to be the JVM shutting down which would exit without waiting for these daemon threads.  We don't really at what point the daemon threads stop or if they stop abruptly since any unfinished work is useless.  So I'm not sure if adding a shutdownHook would do anything different?

On the other hand, if the SparkSession wanted to cancel the running threads with the JVM still running, I think it could do that if it provided it's own ExecutorService.

MLnick commented on mllib/src/main/scala/org/apache/spark/ml/tuning/CrossValidator.scala:
```diff
@@ -100,31 +108,60 @@ class CrossValidator @Since("1.2.0") (@Since("1.4.0") override val uid: String)
+        ThreadUtils.sameThread
+      case n =>
+        ExecutionContext.fromExecutorService(executorServiceFactory(n))
+    }
```
See my other comments about the API for setting executor service.

MLnick commented on mllib/src/main/scala/org/apache/spark/ml/tuning/CrossValidator.scala:
```diff
@@ -100,31 +108,60 @@ class CrossValidator @Since("1.2.0") (@Since("1.4.0") override val uid: String)
-    val metrics = new Array[Double](epm.length)
+
+    // Create execution context, run in serial if numParallelEval is 1
+    val executionContext = $(numParallelEval) match {
```
I think it's ok to make this `implicit val ...` - it makes the `Future` code below clearer and it will be in immediate scope for all the calls so I don't think we need to specify.

BryanCutler replied:
I kind of prefer to explicitly set the execution context.  It does add some clutter, but then it is clear what calls rely on it, just in case some things get moved around in the future.  Do you mind it we leave as is?

MLnick commented on mllib/src/main/scala/org/apache/spark/ml/tuning/CrossValidator.scala:
```diff
@@ -100,31 +108,60 @@ class CrossValidator @Since("1.2.0") (@Since("1.4.0") override val uid: String)
-        i += 1
+
+      // Fit models in a Future with thread-pool size determined by '$numParallelEval'
+      val models = epm.map { paramMap =>
```
If we use the `implicit val` above we can do:

```scala
    val models = epm.map { paramMap =>
      Future { est.fit(trainingDataset, paramMap) }.mapTo[Model[_]]
    }
```

MLnick commented on mllib/src/main/scala/org/apache/spark/ml/tuning/CrossValidator.scala:
```diff
@@ -100,31 +108,60 @@ class CrossValidator @Since("1.2.0") (@Since("1.4.0") override val uid: String)
+        } (executionContext)
       }
+
+      Future.sequence[Model[_], Iterable](models)(implicitly, executionContext).onComplete { _ =>
```
Likewise:

```scala
    Future.sequence[Model[_], Iterable](models).onComplete { _ =>
      trainingDataset.unpersist()
    }
```

MLnick commented on mllib/src/main/scala/org/apache/spark/ml/tuning/CrossValidator.scala:
```diff
@@ -100,31 +108,60 @@ class CrossValidator @Since("1.2.0") (@Since("1.4.0") override val uid: String)
+            val metric = eval.evaluate(model.transform(validationDataset, paramMap))
+            logDebug(s"Got metric $metric for model trained with $paramMap.")
+            metric
+          } (executionContext)
```
Can then remove the `(executionContext)` here and elsewhere

MLnick commented on mllib/src/main/scala/org/apache/spark/ml/tuning/CrossValidator.scala:
```diff
@@ -100,31 +108,60 @@ class CrossValidator @Since("1.2.0") (@Since("1.4.0") override val uid: String)
+
+      // Evaluate models in a Future with thread-pool size determined by '$numParallelEval'
+      val foldMetricFutures = models.zip(epm).map { case (modelFuture, paramMap) =>
+        modelFuture.flatMap { model =>
```
*Note* we could use `for` comprehension here. But I tried it and it doesn't really make it all that much simpler, and the explicit `flatMap` and `Future` here makes it a bit clearer.

MLnick commented on mllib/src/test/scala/org/apache/spark/ml/tuning/CrossValidatorSuite.scala:
```diff
@@ -121,6 +121,33 @@ class CrossValidatorSuite
+
+    val parentSerial = cvSerialModel.bestModel.parent.asInstanceOf[LogisticRegression]
+    val parentParallel = cvParallelModel.bestModel.parent.asInstanceOf[LogisticRegression]
+    assert(parentSerial.getRegParam === parentParallel.getRegParam)
```
Does it make sense to also check `uid` equality here?

BryanCutler replied:
It should probably be done in the test that already runs `checkCopy` on line 62 (at least until we cleanup these basic checks).  I'll take a look at that.

MLnick reviewed (changes requested):
Made a pass on these updates.

I think it overall looks pretty good. Mostly a few clean up comments, and also I think we should look at a simpler API / method for setting a custom execution context.

In terms of `CrossValidator`, I think for now its fine to do parallel execution within each fold. Otherwise it may be very easy to flood a cluster. We can gather user feedback from usage in the wild and always update at a later point.

MLnick commented on mllib/src/main/scala/org/apache/spark/ml/tuning/TrainValidationSplit.scala:
```diff
@@ -87,37 +90,64 @@ class TrainValidationSplit @Since("1.5.0") (@Since("1.5.0") override val uid: St
+      trainingDataset.unpersist()
+    } (executionContext)
+
+    // Evaluate models concurrently, limited by a barrier with '$numParallelEval' permits
```
This comment is stale

MLnick commented on mllib/src/main/scala/org/apache/spark/ml/tuning/CrossValidator.scala:
```diff
@@ -100,31 +108,60 @@ class CrossValidator @Since("1.2.0") (@Since("1.4.0") override val uid: String)
+      }
+
+      // Wait for metrics to be calculated before upersisting validation dataset
+      val foldMetrics = foldMetricFutures.map(ThreadUtils.awaitResult(_, Duration.Inf))
```
Does it make sense to also use `sequence` here?

```scala
    val metrics = (ThreadUtils.awaitResult(
      Future.sequence[Double, Iterable](metricFutures), Duration.Inf)).toArray

```

BryanCutler replied:
I thought about that, but since it's a blocking call anyway, it will still be bound by the longest running thread.

MLnick replied:
Sure, not a big deal either way

MLnick commented on mllib/src/main/scala/org/apache/spark/ml/tuning/CrossValidator.scala:
```diff
@@ -100,31 +108,60 @@ class CrossValidator @Since("1.2.0") (@Since("1.4.0") override val uid: String)
+        } (executionContext)
+      }
+
+      // Wait for metrics to be calculated before upersisting validation dataset
```
Typo: `upersisting` -> `unpersisting`

BryanCutler committed:
made doc changes

BryanCutler committed:
changed ExecutorService factory to a trait to be compatible with Java

SparkQA commented:
**[Test build #75716 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75716/testReport)** for PR 16774 at commit [`5e8a086`](https://github.com/apache/spark/commit/5e8a0869dcefaa5febf6cc354a7840225268acf9).

BryanCutler commented:
Thanks for the review @MLnick!  I changed `setExecutorService` to use a trait instead of just a function, which can be implemented in Java.  Works the same, but does add the public trait if that is ok.

BryanCutler commented on examples/src/main/scala/org/apache/spark/examples/ml/ModelSelectionViaTrainValidationSplitExample.scala:
```diff
@@ -65,6 +65,8 @@ object ModelSelectionViaTrainValidationSplitExample {
       // 80% of the data will be used for training and the remaining 20% for validation.
       .setTrainRatio(0.8)
+      // Evaluate up to 2 parameter settings in parallel
+      .setNumParallelEval(2)
```
TODO: I should probably set this in Java too, to be consistent

SparkQA commented:
**[Test build #75716 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75716/testReport)** for PR 16774 at commit [`5e8a086`](https://github.com/apache/spark/commit/5e8a0869dcefaa5febf6cc354a7840225268acf9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `trait ExecutorServiceFactory `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75716/
Test PASSed.

MLnick commented:
@BryanCutler getting back to this after a long delay now that `2.2` is about ready. Sorry about that!

You mentioned the approach above for creating the `ExecutorService` being useful because it is instantiated at the start of `fit` (which is good) and then cleaned up. However, I don't see any explicit cleanup of it happening and if we create a daemon thread pool won't that just hang around?

Overall, I prefer the following semantics:
1. If no custom context is set, we create the default one within `fit` and clean it up at the end of `fit`.
2. For most use cases, even on shared resource cluster / notebook setups, this should suffice as we're creating a dedicated threadpool that only lives while `fit` is called. For custom contexts, that is really an expert setting. I would prefer to keep the API very simple as I mentioned above: `def setExecutorService(executorService: ExecutorService): Unit`, rather than introducing a new trait for this purpose. I think the overhead of creating a (typically relatively small) threadpool non-lazily before instantiating the validator class is not significant. Experts can then decide exactly what they want to do, and when / how to clean up their threadpool (e.g. they may want to keep it around for some reason).

BryanCutler commented:
Thanks for getting back to this @MLnick.  The only difference with daemon threads is they won't prevent the JVM from shutting down - which is what we would want in any case.  The name is a bit misleading, but the JVM will not shutdown until all non-daemon threads have finished, so if the main thread is done and only daemon threads are still running, then the JVM will shutdown immediately.  This is what we would want here because if the thread invoking CrossValidation is interrupted, then we do not want to wait for all the worker threads to finish, we want them to stop immediately.  It doesn't matter if the work is incomplete because unless CrossValidation has run it's course the results are not really useful, so graceful shutdown is not really needed here.

By cleanup, I mean the threadpool should be scoped to the `fit` method, not the class instance, so any resources wouldn't be hanging around after calling `fit` once the threadpool went out of scope - you don't really need to cleanup explicitly.  We can still do this without using the factory.

I guess the reason for me using the factory pattern was attempting to be as flexible as possible and still allow using the `numParallelEval` param with a custom executor.  My thinking was thatcompletley maybe  `numParallelEval` is a user level param and the `ExecutorService` would be configured on some higher level.  Since none of this is in use today and it could be a completely different when it does, it would make sense to just make the API simple like you mentioned until it can be better hashed out.  So I'll redo this with using `def setExecutorService(executorService: ExecutorService): Unit` - sorry for the long-winded response!

BryanCutler committed:
Merge remote-tracking branch 'upstream/master' into parallel-model-eval-SPARK-19357

BryanCutler committed:
Changed ExecutorService to be set explicitly instead of factory

SparkQA commented:
**[Test build #78018 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78018/testReport)** for PR 16774 at commit [`ad8a870`](https://github.com/apache/spark/commit/ad8a8700306869722cb19b6b16c030b343a209c0).

SparkQA commented:
**[Test build #78018 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78018/testReport)** for PR 16774 at commit [`ad8a870`](https://github.com/apache/spark/commit/ad8a8700306869722cb19b6b16c030b343a209c0).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78018/
Test FAILed.

BryanCutler commented:
retest this please

SparkQA commented:
**[Test build #78114 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78114/testReport)** for PR 16774 at commit [`ad8a870`](https://github.com/apache/spark/commit/ad8a8700306869722cb19b6b16c030b343a209c0).

SparkQA commented:
**[Test build #78114 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78114/testReport)** for PR 16774 at commit [`ad8a870`](https://github.com/apache/spark/commit/ad8a8700306869722cb19b6b16c030b343a209c0).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78114/
Test PASSed.

WeichenXu123 reviewed (commented):
I think we can optimize to reduce the memory cost to be O(NumParallelEval * model_size).
Similar to what is done in #18733

hhbyyh commented:
I'm confused by your suggestions here and in  #18733. 

I don't think it's appropriate to just "include" a similar work originated from another PR, and suggest another PR to suspend.

BryanCutler commented:
@WeichenXu123 , if #18733 is ready to be merged, then it should not be put on hold because of this PR.  We also want to make sure to work collaboratively if there is overlap, and respect other peoples work.

I'll just point out that this is using `Future[Model]`s and the Model would only be in memory once the future is completed.  Then after it is evaluated it is free to be cleaned up by gc.

WeichenXu123 commented:
@BryanCutler You are right. Once `Future` complete the model can be cleaned by GC. So the memory cost of the code has been optimized already. I didn't look at the code carefully a few days ago.
Thanks for your careful work for this. I will take a deeper look later. ^_^

BryanCutler commented:
Thanks for taking a look @WeichenXu123!

SparkQA commented:
**[Test build #3896 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3896/testReport)** for PR 16774 at commit [`ad8a870`](https://github.com/apache/spark/commit/ad8a8700306869722cb19b6b16c030b343a209c0).

MLnick commented:
@BryanCutler can we pick the `HasParallel` trait from #18281 into this PR?

Then this PR won't be blocked, and when #18281 is ready it should still be cleanly mergeable.

WeichenXu123 commented:
@BryanCutler @MLnick I agree pick `HasParallel` into this PR because the `trait` has very little code. Another feature is pending on this PR. So we hope this get merged soon! cc @jkbradley

BryanCutler committed:
added HasParallelism trait

BryanCutler committed:
Updated to use Trait HasParallelsim

SparkQA commented:
**[Test build #81046 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81046/testReport)** for PR 16774 at commit [`658aacb`](https://github.com/apache/spark/commit/658aacb8f4e8e7937f2bdd81074a54a60dee2037).

BryanCutler committed:
fixed up docs

SparkQA commented:
**[Test build #81047 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81047/testReport)** for PR 16774 at commit [`2c73b0b`](https://github.com/apache/spark/commit/2c73b0bce559236bbe32f1b5237e2fb17ef5af58).

SparkQA commented:
**[Test build #81046 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81046/testReport)** for PR 16774 at commit [`658aacb`](https://github.com/apache/spark/commit/658aacb8f4e8e7937f2bdd81074a54a60dee2037).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81046/
Test FAILed.

SparkQA commented:
**[Test build #81047 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81047/testReport)** for PR 16774 at commit [`2c73b0b`](https://github.com/apache/spark/commit/2c73b0bce559236bbe32f1b5237e2fb17ef5af58).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81047/
Test PASSed.

BryanCutler commented:
@MLnick @WeichenXu123 I updated to use the trait `HasParallel` and fixed up some of the docs, please take another look, thanks!

BryanCutler commented:
ping @MLnick , does this look ok to merge?

WeichenXu123 commented on mllib/src/test/scala/org/apache/spark/ml/tuning/CrossValidatorSuite.scala:
```diff
@@ -120,6 +120,33 @@ class CrossValidatorSuite
+      .setEstimatorParamMaps(lrParamMaps)
+      .setEvaluator(eval)
+      .setNumFolds(2)
+      .setParallelism(1)
```
add `.setSeed(XXX)`

BryanCutler replied:
So the seed param here is fixed by default and doesn't need to be set to ensure consistent results.  I think that's why it's not set in the other tests in this suite.  I'm not a fan of this behavior and I think it's better to explicitly set in tests, but then we should probably be consistent and set elsewhere too.  What are your thoughts on this @MLnick ?

WeichenXu123 replied:
OK I agree.

MLnick replied:
Yeah seed defaults to a hash of the class name. There has been debate over this (see [SPARK-16832](https://issues.apache.org/jira/browse/SPARK-16832)). Personally I also don't like that behavior, but for now that's what it is.

WeichenXu123 reviewed (commented):
LGTM except a minor comment, thanks!

felixcheung commented on mllib/src/test/scala/org/apache/spark/ml/tuning/CrossValidatorSuite.scala:
```diff
@@ -120,6 +120,33 @@ class CrossValidatorSuite
+      .setNumFolds(2)
+      .setParallelism(1)
+    val cvSerialModel = cv.fit(dataset)
+    cv.setParallelism(2)
```
how do we validate setParallelism is parallelizing?

BryanCutler replied:
It's a little difficult to do this in a unit test without making it flaky.  I have run tests manually and verified it is working by both the expected speedup in timing and that the expected number of tasks are run concurrently.  I can post some results if that would help.

MLnick replied:
@BryanCutler may be worth posting the result to the JIRA for posterity.

WeichenXu123 commented on mllib/src/main/scala/org/apache/spark/ml/tuning/CrossValidator.scala:
```diff
@@ -100,31 +113,53 @@ class CrossValidator @Since("1.2.0") (@Since("1.4.0") override val uid: String)
-        i += 1
+
+      // Fit models in a Future for training in parallel
+      val models = epm.map { paramMap =>
```
I think use var name `modelFutures` instead of `models` will be more clear.

WeichenXu123 commented on mllib/src/main/scala/org/apache/spark/ml/tuning/CrossValidator.scala:
```diff
@@ -100,31 +113,53 @@ class CrossValidator @Since("1.2.0") (@Since("1.4.0") override val uid: String)
+    }.transpose.map(_.sum)
+
+    // Calculate average metric over all splits
     f2jBLAS.dscal(numModels, 1.0 / $(numFolds), metrics, 1)
```
I don't like this line code because it use low-level api which make the code difficult to read.
And here is not bottleneck.
So I think we can simply use:
```
val metrics = ...
   ....
   .transpose.map(_.sum / $(numFolds))
```
instead.
What do you think about it ?

MLnick replied:
I must say I wondered why the author of this bothered with a BLAS call to do the computation... even if there were a few 1000s param combinations it would still not be significant overhead. +1

BryanCutler replied:
Yeah, I agree.  I'll go ahead and make that change here.

WeichenXu123 commented on mllib/src/main/scala/org/apache/spark/ml/tuning/TrainValidationSplit.scala:
```diff
@@ -87,37 +91,63 @@ class TrainValidationSplit @Since("1.5.0") (@Since("1.5.0") override val uid: St
-      logDebug(s"Got metric $metric for model trained with ${epm(i)}.")
-      metrics(i) += metric
-      i += 1
+    val models = epm.map { paramMap =>
```
`models` ==> `modelFutures`

WeichenXu123 reviewed (commented):
LGTM except minor comments. Thanks!

mgaido91 commented on mllib/src/main/scala/org/apache/spark/ml/tuning/CrossValidator.scala:
```diff
@@ -100,31 +113,53 @@ class CrossValidator @Since("1.2.0") (@Since("1.4.0") override val uid: String)
-    val metrics = new Array[Double](epm.length)
+
+    // Create execution context based on $(parallelism)
+    val executionContext = getExecutionContext
```
In the corresponding PR for PySpark implementation the number of threads is limited by the number of models to be trained (https://github.com/WeichenXu123/spark/blob/be2f3d0ec50db4730c9e3f9a813a4eb96889f5b6/python/pyspark/ml/tuning.py#L261). We might do that for instance by overriding the `getParallelism` method. What do you think about this?

MLnick reviewed (approved):
LGTM pending the minor outstanding comments from @WeichenXu123

jkbradley commented:
I think https://github.com/apache/spark/pull/19110 is ready to merge now.  @BryanCutler @WeichenXu123 do you have a preference for which gets merged first?

BryanCutler committed:
removed blas calculation for CrossValidator metric calc, was not necessary

SparkQA commented:
**[Test build #81417 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81417/testReport)** for PR 16774 at commit [`7a8221b`](https://github.com/apache/spark/commit/7a8221ba939da026a9818299cf2b897ea81766a5).

BryanCutler commented:
If you wouldn't mind, please merge this first (once passes Jenkins).  I had already updated this with the HasParallelism trait after waiting on OneVsRest for a while. Thanks!

SparkQA commented:
**[Test build #81417 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81417/testReport)** for PR 16774 at commit [`7a8221b`](https://github.com/apache/spark/commit/7a8221ba939da026a9818299cf2b897ea81766a5).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81417/
Test PASSed.

BryanCutler commented:
This should be ready to merge @jkbradley @MLnick

MLnick commented:
LGTM. Merged this to master. Thanks @BryanCutler and everyone for reviewing!

(in this PR the doc in the trait is a little more detailed which is slightly better IMO).

WeichenXu123 commented:
@BryanCutler @MLnick I found a bug in this PR: after save estimator (CV or TVS) and then load again, the "Parallelism" setting will be lost. But I fix this in #19208 by the way.

jkbradley commented:
@WeichenXu123 Thanks for finding that bug!  Can you please separate out your bugfix?  It's good to get fixes in, rather than attaching them to PRs which may require discussion, so that we make sure that bugs don't slip into the next release.

BryanCutler commented:
@WeichenXu123 , it would be great if you could separate out the bugfix. I looked in #19208 but couldn't find what you were referring to.

WeichenXu123 commented:
OK. I will separate a PR. Thanks!
